### DISCUSSION TITLE
Problem about correcting force constant after SCPH calculation using ALAMODE 1.3.0

### DISCUSSION THREAD

**Discussion by XiangXing96** [ttadano/alamode#71] [Q&A]
Dear Dr. Tadano,

I use ALAMODE to calculate the 2nd(harmonic)-, 3rd- and 4th- order force constant with the force and trajectories from lammps and then conduct scph calculations. There are no warning appears during the whole process. However, when I use scph_dfc2 file to correct the 2nd-order force constant, warnings appear take place like
Warning: The following force constant doesn't exist in the original file:
   -4   -4   -1    0    0    1    1    0.000684271
Warning: The following force constant doesn't exist in the original file:
   -4   -4   -1    0    0    1    2   -0.000481604
Warning: The following force constant doesn't exist in the original file:
   -4   -4   -1    0    1    1    0    0.000684271

Based on your previous suggestion for the problem, I check the numerical digits of lattice vectors and fractional coordinates as well as KMESH_INTERPOLATE. The numerical digits are 15 and the KMESH_INTERPOLATE is consistent with the supercell size. Additionally, when I conduct the whole calculation with 222 supercell, the problem do not take place.

Could you give me some suggestions about the problem? For convenience, I attach the harmonic force constant input (run.harm), high-order force constant input (run.high-order-fitting) and scph input (scph.in).

**run.harm:**
![image](https://user-images.githubusercontent.com/50902454/162760955-ab140e1e-13df-43aa-9b7e-de352962aa59.png)

**high-order force constant input:**
![image](https://user-images.githubusercontent.com/50902454/162761282-6260f514-b567-40fe-a629-035834217435.png)

**scph.in:**
![image](https://user-images.githubusercontent.com/50902454/162761502-945fed7b-4bde-470f-8703-a5c25b570304.png)

**Comment by ttadano** [comment]
Hi,

All input files look good except for the numbers in the &position field like 0.124999999. 

Could you round all of these numbers to a fraction, like 0.125, and see if the problem persists.

**Reply by XiangXing96** [reply]
Thanks for your reply. I have rounded all of these numbers, but the problem still persists. Can the problem be caused by the accuracy of fourth-order force constant (due to the cutoff, the number of dataset, etc.)?

**Reply by ttadano** [reply]
Thank you for your prompt check. This problem should not be caused by the accuracy of the force constants. It indicates a possible breakdown of the force constant symmetry. To investigate the origin more, I need the input files. If you don't mind, please send the input files directly to me.

**Reply by ttadano** [reply]
Hi @XiangXing96. After getting inputs directly from you, I've investigated the reason of the warning message and identified it. 

**Quick solution**
Please set `FC_ZERO_THR = 1.0e-15` in the `&general` field of ALM when computing harmonic force constants (`harmonic.xml`) and use this harmonic FCSXML when generating an effective harmonic FCSXML using `dfc2`. `FC_ZERO_THR` is available in the develop branch.

**Detailed explanation**
In `alm`, after the force constants are fitted to the data, the code judges if the absolute value of the force constant is larger than 1.0e-12. If it is smaller than this threshold, the code does NOT print that force constant in `PREFIX.xml` (e.g. `harmonic.xml`) because such a small force constant does not affect phonon calculation result. This treatment is OK as long as the forces are computed using a DFT code.

When the forces are computed using _a model potential_ (e.g. classical or machine-learning FFs) as you did, the harmonic force constants may become almost zero (<1.0e-12) when the distance between two atoms are larger than a model-specific cutoff radius. The key point here is that these terms become almost zero **NOT** because of symmetry constraints. So, these terms can take a non-zero value when the model-specific cutoff raius is increased.

In the SCPH mode of `anphon`, the symmetry is fully taken into account, but the code does not know if the force constant are generated by a model potential or not. Hence, the anharmonic correction to the force constants (saved in `PREFIX.scph_dfc2`) can take a non-zero value even when the distance between the two involving atoms are larger than the cutoff distance.  However, if the corresponding term does not exist in `harmonic.xml`, an warning message is raised when creating renormalized harmonic force constants file using `dfc2`.

To suppress this warning, we have added a new tag `FC_ZERO_THR` for `alm`. This tag defines the threshold of force constants to be regarded as zero. The default value is 1.0e-12, but a much smaller value is recommended for a model potential with a cutoff radius, particularly when using for `dfc2`. I've confirmed that the warning you observed can be suppressed with `FC_ZERO_THR=1.0e-15`.

**Reply by XiangXing96** [reply]
Hi, Dr. Tadano,
I get it. Thanks very much for your helps.

**Reply by XiangXing96** [reply]
Hi, Dr. Tadano,
Sorry for bothering you again. 
Recently, I check the case again but the warning still takes place even with FC_ZERO_THR=1.0e-15 ([make.log](https://github.com/ttadano/alamode/files/8695355/make.log)). Therefore, I want to check the space group of structure recognized by alm and anphon. I find Space group information in the log file of alm, while that does not appear in the log file of anphon. Is there a way to find that? Additionally, I want to know whether the "Atomic positions in the primitive cell (fractional):" in the log file of anphon ([scph.log](https://github.com/ttadano/alamode/files/8695347/scph.log)) is the atomic position in reciprocal space. I find they are different from those listed in FC2XML, just like . Could you give me a favor? Finally, I want to use a finite cutoff for the harmonic force constant, which is the cutoff of the model potential. Will that make the warning more serious?
Thank you in advance.

**Reply by ttadano** [reply]
> Recently, I check the case again but the warning still takes place even with FC_ZERO_THR=1.0e-15 ([make.log](https://github.com/ttadano/alamode/files/8695355/make.log)).

Sorry, I forgot to comment out some debug print lines in the previous commit. The outputs shown in your make.log are not warning, so your calculation seems to have finished properly.

> I find Space group information in the log file of alm, while that does not appear in the log file of anphon. Is there a way to find that?

The anphon code does not print the space group symbol. It just prints the number of space group (SG) operations. 
If (# of SG operations in anphon) != (# of SG operations in alm) * (# of primitive cells in the supercell) does not hold, it indicates that anphon fails to detect the SG symmetry. If the above equation holds, the calculation should be OK.

> Additionally, I want to know whether the "Atomic positions in the primitive cell (fractional):" in the log file of anphon ([scph.log](https://github.com/ttadano/alamode/files/8695347/scph.log)) is the atomic position in reciprocal space. I find they are different from those listed in FC2XML, just like . Could you give me a favor?

Yes, they are fractional coordinates of atoms in a primitive cell detected by anphon. Where did you see the difference?

> Finally, I want to use a finite cutoff for the harmonic force constant, which is the cutoff of the model potential. Will that make the warning more serious?

I don't know it if it would be more serious or not. In general, if you see a warning, it means that the anharmonic correction to the 2nd-order IFCs has longer-range components than the original 2nd-order IFCs. This situation may be a little strange because the anharmonic interaction is usually more short-ranged than the harmonic interaction. My suggestion is to make `KMESH_INTERPOLATE` smaller until the warning disappears.

**Reply by XiangXing96** [reply]
Hi, Dr. Tadano,
Thanks for your reply.

For the first question:
In make.log, some elements in the last column are still pretty large and thus there seems still some problems in this calculation. 

For the second question:
Based on your replies, I check the (# of SG operations in anphon) != (# of SG operations in alm) * (# of primitive cells in the supercell). In my calculation, # of SG operations in anphon, alm and  # of primitive cells in the supercell are 4, 108 and 27. Additionally, I give a primitive cell into the anphon and alm and then output the symmetry information, respectively. I find that the symmetry information in alm.log is almost the same to that in the initial cif file, while that in anphon.log is different. Can the difference lead to the warning in make.log?
alm.log
![image](https://user-images.githubusercontent.com/50902454/168558562-32f8b950-0451-4cbc-9f2f-77bbe39668c4.png)
anphon.log
![image](https://user-images.githubusercontent.com/50902454/168558652-934bebe8-66b6-4998-8be8-dfa767897caa.png)

For the third question:
Ok, I will have a try to decrease the KMESH_INTERPOLATE.

**Reply by ttadano** [reply]
> For the first question:
> In make.log, some elements in the last column are still pretty large and thus there seems still some problems in this calculation.

If you haven't pull the latest update on the develop branch, please do so. I think the warning messages will disappear with the new version.

> Based on your replies, I check the (# of SG operations in anphon) != (# of SG operations in alm) * (# of primitive cells in the supercell). In my calculation, # of SG operations in anphon, alm and # of primitive cells in the supercell are 4, 108 and 27. 

Do you mean  (# of SG operations in anphon) **=** (# of SG operations in alm) * (# of primitive cells in the supercell)?

> Additionally, I give a primitive cell into the anphon and alm and then output the symmetry information, respectively. I find that the symmetry information in alm.log is almost the same to that in the initial cif file, while that in anphon.log is different. Can the difference lead to the warning in make.log?

Maybe, that is not problematic because +0.5 and -0.5 are equivalent.
